lib/split/experiment.rb

```diff
@@ -325,11 +323,154 @@ def goals_key
     end
 
-    def finished_key(goal=nil)
+    def finished?(split_id, goal = nil)
+      @finished ||= {}
+
+      goalkey = goal || "nogoal"
+
+      if !@finished[split_id].nil? && !@finished[split_id][goalkey].nil?
+        return @finished[split_id][goalkey]
+      else
+        key = "#{self.key}:finished"
+        key << ":#{goal}" if goal
+        value = Split.redis.with do |conn|
+          conn.sismember(key, split_id)
+        end
+
+        @finished[split_id] ||= {}
+        @finished[split_id][goalkey] = value
+
+        return value
+      end
+    end
+
+    def cache_finished!(split_id, value, goal = nil)
+      @finished ||= {}
+
+      @finished[split_id] ||= {}
+
       if goal
-        "#{key}:#{goal}:finished"
+        @finished[split_id][goal] = value
+      else
+        @finished[split_id]["nogoal"] = value
+      end
+    end
+
+    def self.preload_participating!(experiments, split_ids)
+      experiments = Array(experiments)
+      split_ids = Array(split_ids)
+
+      experiments_metadata = []
+      experiments.each do |experiment|
+        split_ids.each do |split_id|
+          key = "#{experiment.key}:participants"
+
+          experiments_metadata << {
+              :key => key,
+              :experiment => experiment,
+              :split_id => split_id
+          }
+        end
+      end
+
+      if experiments_metadata.present?
+        results = Split.redis.with do |conn|
+          conn.pipelined do
+            experiments_metadata.each do |metadata|
+              conn.sismember metadata[:key], metadata[:split_id]
+            end
+          end
+        end
+
+        results.each_with_index do |result, index|
+          metadata = experiments_metadata[index]
+          metadata[:experiment].cache_participating!(metadata[:split_id], result)
+        end
+      end
+    end
+
+    def self.preload_finished!(experiments, goals, split_ids)
+      experiments = Array(experiments)
+      split_ids = Array(split_ids)
+
+      experiments_metadata = []
+      goals.each do |goal|
+        experiments.each do |experiment|
+          split_ids.each do |split_id|
+            key = "#{experiment.key}:finished:#{goal}"
+
+            experiments_metadata << {
+                :key => key,
+                :experiment => experiment,
+                :goal => goal,
+                :split_id => split_id
+            }
+          end
+        end
+      end
+
+      if experiments_metadata.present?
+        results = Split.redis.with do |conn|
+          conn.pipelined do
+            experiments_metadata.each do |metadata|
+              conn.sismember metadata[:key], metadata[:split_id]
+            end
+          end
+        end
+
+        results.each_with_index do |result, index|
+          metadata = experiments_metadata[index]
+          metadata[:experiment].cache_finished!(metadata[:split_id], result, metadata[:goal])
+        end
+      end
+    end
+
+    def finish!(split_id, goal = nil)
```

Comment:
not necessary, but looks like i used nogoal under `delete_finished`. that's addressed now. good catch.